File: PllhaC++/PilhaEstatica.cpp
//
// Created by Camila on 18/08/2020.
//

#include "PilhaEstatica.h"

PilhaEstatica::PilhaEstatica(int max)
: dados(new int[max]), topo(-1), max(max){
}

bool PilhaEstatica::vazia() const {
    return topo == -1;
}

void PilhaEstatica::empilha(int valor) {
    topo = topo +1;
    dados[topo] = valor;
}

int PilhaEstatica::desempilha() {
    int dado = dados[topo];
    topo = topo -1;
    return dado;
}

PilhaEstatica::~PilhaEstatica() {
    delete [] dados;
}

//PilhaEstatica::PilhaEstatica(const PilhaEstatica &outra) {
//    this->topo = outra.topo;
//    this->max = outra.max;
//
//    this->dados = new int[max];
//    for (int i = 0; i < max; i++) {
//        this->dados[i] = outra.dados[i];
//    }
//}

//PilhaEstatica &PilhaEstatica::operator=(const PilhaEstatica &outra) {
//    if(&outra == this) return *this;
//
//    this->topo = outra.topo;
//    this->max = outra.max;
//
//    delete [] dados;
//    this->dados = new int[max];
//    for (int i = 0; i < max; i++) {
//        this->dados[i] = outra.dados[i];
//    }
//    return *this;
//}

File: Pilha/src/PilhaEstatica.java
public class PilhaEstatica implements Pilha,Cloneable
{

    int topo;
    int[] dados;

    public PilhaEstatica(int max)
    {
        topo = -1;
        dados = new int[max];
    }

    public PilhaEstatica(PilhaEstatica outra)
    {
        this.topo = outra.topo;
        this.dados = new int[outra.dados.length];
        for (int i = 0; i < outra.dados.length; i++)
        {
            this.dados[i] = outra.dados[i];
        }
    }

    @Override
    public boolean vazia()
    {
        return topo == -1;
    }

    @Override
    public void empilha(int valor)
    {
            topo++;
            dados[topo] = valor;
    }

    @Override
    public int desempilha()
    {
        if (vazia())
        {
            return -1;
        }
        else
        {
            int dado = dados[topo];
            topo--;
            return dado;
        }
    }

    protected PilhaEstatica clone(){
        try {
            PilhaEstatica clone = (PilhaEstatica) super.clone();
            clone.dados = new int[dados.length];
            for (int i = 0; i < dados.length; i++){
                clone.dados[i] = dados[i];
            }
            return clone;
        } catch (CloneNotSupportedException e) {
            return null;
        }
    }
}
